src/gui/dialogs/statistics_dialog.cpp

```diff
@@ -133,34 +151,169 @@ void statistics_dialog::add_damage_row(
+		return str.str();
+	};
+	item["label"] = damage_str(damage, expected);
 	data.emplace("damage_overall", item);
 
-	str.str("");
+	const auto& percent_str = [](long long dsx, long long expected) {
+		std::ostringstream str;
+		str << (((dsx < 0) ^ (expected < 0)) ? "" : "+")
+			<< (expected == 0 ? 0 : 100 * dsx / expected) << '%';
+		return str.str();
+	};
+	item["label"] = percent_str(dsa, expected);
+	data.emplace("overall_percent", item);
 
 	if(show_this_turn) {
-		const long long turn_shifted = ((turn_expected * 20) + shift) / (2 * shift);
-		str << turn_damage << " / "
-			<< static_cast<double>(turn_shifted) * 0.1
-			<< "    " // TODO: should probably make this two columns
-			<< (((dst < 0) ^ (turn_expected < 0)) ? "" : "+")
-			<< (turn_expected == 0 ? 0 : 100 * dst / turn_expected) << '%';
-
-		item["label"] = str.str();
+		item["label"] = damage_str(turn_damage, turn_expected);
 		data.emplace("damage_this_turn", item);
+
+		item["label"] = percent_str(dst, turn_expected);
+		data.emplace("this_turn_percent", item);
 	}
 
 	damage_list.add_row(data);
 }
 
+// Custom type to allow tally() to return two values.
+struct hitrate_table_element {
+	// The string with <actual number of hits>/<expected number of hits>
+	std::string hitrate_str;
+	// The string with the a priori probability of that result
+	std::string percentage_str;
+};
+
+// Return the strings to use in the "Hits" table, showing actual and expected number of hits.
+static hitrate_table_element tally(const statistics::stats::hitrate_map& by_cth, const bool more_is_better)
+{
+	unsigned int overall_hits = 0;
+	double expected_hits = 0;
+	unsigned int overall_strikes = 0;
+
+	for(const auto& i : by_cth) {
+		int cth = i.first;
+		overall_hits += i.second.hits;
+		expected_hits += (cth * 0.01) * i.second.strikes;
+		overall_strikes += i.second.strikes;
+	}
+
+	std::ostringstream str, str2;
+	str << overall_hits << " / " << expected_hits;
+
+	// Compute the a priori probability of this actual result, by simulating many attacks against a single defender.
+	{
+		config defender_cfg(
+			"id", "statistics_dialog_dummy_defender",
+			"hide_help", true,
+			"do_not_list", true,
+			"hitpoints", overall_strikes
+		);
+		unit_type defender_type(defender_cfg);
+		unit_types.build_unit_type(defender_type, unit_type::BUILD_STATUS::FULL);
+
+		battle_context_unit_stats defender_bc(&defender_type, nullptr, false, nullptr, nullptr, 0 /* not used */);
+		std::unique_ptr<combatant> current_defender(new combatant(defender_bc));
+
+		for(const auto& i : by_cth) {
+			int cth = i.first;
+			config attacker_cfg(
+				"id", "statistics_dialog_dummy_attacker" + std::to_string(cth),
+				"hide_help", true,
+				"do_not_list", true,
+				"hitpoints", 1
+			);
+			unit_type attacker_type(attacker_cfg);
+			unit_types.build_unit_type(attacker_type, unit_type::BUILD_STATUS::FULL);
+
+			attack_ptr attack = std::make_shared<attack_type>(config(
```

Comment:
Just use `auto` here instead of `attack_ptr` since you're using `make_shared`.

Reply:
e9d7b248c6c6b94ff340d831549638cef51b08f4